src/Illuminate/View/Concerns/CachesComponents.php

```diff
@@ -0,0 +1,126 @@
+namespace Illuminate\View\Concerns;
+
+use Illuminate\Container\Container;
+use Illuminate\Contracts\Support\Htmlable;
+use Illuminate\Contracts\View\CanBeCached;
+use Illuminate\Contracts\View\View;
+use Illuminate\Support\Facades\Cache;
+
+trait CachesComponents
+{
+    /**
+     * The cache controlling attributes for this component.
+     *
+     * @return array
+     */
+    protected function cacheAttributes()
+    {
+        return [
+            'x-cache',
+            'x-cache-key',
+            'x-cache-ttl',
+        ];
+    }
+
+    /**
+     * Determines if the component can be cached.
+     *
+     * @return bool
+     */
+    protected function canBeCached()
+    {
+        return $this instanceof CanBeCached;
+    }
+
+    /**
+     * Determines if the component should be cached.
+     *
+     * @return bool
+     */
+    protected function shouldBeCached()
+    {
+        return $this->attributes->get('x-cache', false) && $this->canBeCached();
+    }
+
+    /**
+     * The key used to cache the component.
+     *
+     * @param  array  $slots
+     * @return bool
+     */
+    protected function cacheKey(array $slots)
+    {
+        return $this->attributes->get('x-cache-key', static::class.sha1($this->attributes->except($this->cacheAttributes())->toHtml().serialize($slots)));
+    }
+
+    /**
+     * The cache TTL to use for the component.
+     *
+     * @return mixed
+     */
+    protected function cacheTtl()
+    {
+        return $this->attributes->get('x-cache-ttl', now()->addHour());
+    }
+
+    /**
+     * Render the resolved component view so that it can be cached.
+     *
+     * @param  \Illuminate\Contracts\View\View|\Illuminate\Contracts\Support\Htmlable|string $view
+     * @param  array  $data
+     * @return string
+     *
+     * @throws \Illuminate\Contracts\Container\BindingResolutionException
+     */
+    protected function renderResolvedView($view, $data)
+    {
+        $data['attributes'] = $data['attributes']->filter(function ($value, $attribute) {
+            return ! in_array($attribute, $this->cacheAttributes());
+        });
+
+        if ($view instanceof View) {
+            return $view->with($data)->render();
+        }
+
+        if ($view instanceof Htmlable) {
+            return $view->toHtml();
+        }
+
+        return Container::getInstance()->make('view')->make($view, $data)->render();
+    }
+
+    /**
+     * Cache the resolved component view and return a Htmlable class that will
+     * return the cached result.
+     *
+     * @param  \Illuminate\Contracts\View\View|\Illuminate\Contracts\Support\Htmlable|string  $view
+     * @param  array  $data
+     * @return \Illuminate\Contracts\Support\Htmlable
+     */
+    protected function cacheResolvedView($view, $data)
+    {
+        if (! $this->shouldBeCached()) {
+            return $view;
+        }
+
+        $html = Cache::remember($this->cacheKey($data['__laravel_slots']), $this->cacheTtl(), function () use ($view, $data) {
```

Comment:
I wasn't sure if the facade can be used here, or if this should be pulled out of the container.
Also - should the cache be customizable? Could be another attribute, like `x-cache-driver="redis"`

Reply:
*If* we use the Cache component here it indeed will need to be pulled from the container. I don't think we should make that configurable through attributes (or at all).